### ISSUE TITLE
PHASE 3: which model is best for the task

### ISSUE THREAD

**Issue by brandy333-umbre** [brandy333-umbre/medical-no-show-appointments#2]
Script:
"""
PHASE 3 — Model Comparison & Robustness
========================================
Compares Logistic Regression (baseline), Random Forest, and Gradient Boosting.

Evaluates:
  - ROC-AUC and PR-AUC via Stratified K-Fold cross-validation
  - Performance variance (stability)
  - Calibration quality (Brier score + calibration curves)
  - Overfitting check (train vs. test gap)
"""

from __future__ import annotations

import warnings
import numpy as np
import pandas as pd

from sklearn.model_selection import StratifiedKFold
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import (
    roc_auc_score,
    average_precision_score,
    brier_score_loss,
)
from sklearn.calibration import calibration_curve

import sys
import os

# Ensure src/ is on the path so train_model can be imported from anywhere
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from train_model import load_and_clean, DATA_PATH

# Resolve DATA_PATH relative to project root (one level above src/)
PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
RESOLVED_DATA_PATH = os.path.join(PROJECT_ROOT, DATA_PATH)

warnings.filterwarnings("ignore", category=UserWarning)

N_FOLDS = 5
RANDOM_STATE = 42

# ------------------------------------------------------------------
# Preprocessing (mirrors train_model.py but built fresh per fold)
# ------------------------------------------------------------------
def _build_preprocessor(X: pd.DataFrame) -> ColumnTransformer:
    num_cols = X.select_dtypes(include=["int64", "float64"]).columns.tolist()
    cat_cols = X.select_dtypes(include=["object", "string", "bool"]).columns.tolist()

    numeric_transformer = Pipeline([
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler()),
    ])
    categorical_transformer = Pipeline([
        ("imputer", SimpleImputer(strategy="most_frequent")),
        ("onehot", OneHotEncoder(handle_unknown="ignore")),
    ])

    return ColumnTransformer(
        transformers=[
            ("num", numeric_transformer, num_cols),
            ("cat", categorical_transformer, cat_cols),
        ],
        remainder="drop",
    )

# ------------------------------------------------------------------
# Model definitions
# ------------------------------------------------------------------
def get_models() -> dict[str, object]:
    return {
        "Logistic Regression": LogisticRegression(
            max_iter=2000,
            class_weight="balanced",
            solver="liblinear",
        ),
        "Random Forest": RandomForestClassifier(
            n_estimators=300,
            max_depth=12,
            min_samples_leaf=20,
            class_weight="balanced",
            random_state=RANDOM_STATE,
            n_jobs=-1,
        ),
        "Gradient Boosting": GradientBoostingClassifier(
            n_estimators=200,
            max_depth=4,
            learning_rate=0.1,
            subsample=0.8,
            random_state=RANDOM_STATE,
        ),
    }

def _build_pipeline(preprocessor: ColumnTransformer, clf: object) -> Pipeline:
    return Pipeline([
        ("preprocess", preprocessor),
        ("clf", clf),
    ])

# ------------------------------------------------------------------
# Cross-validation loop
# ------------------------------------------------------------------
def cross_validate_models(
    X: pd.DataFrame,
    y: np.ndarray,
) -> pd.DataFrame:
    """Run Stratified K-Fold CV and collect per-fold metrics for each model."""
    skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=RANDOM_STATE)
    models = get_models()
    records: list[dict] = []

    for fold_idx, (train_idx, test_idx) in enumerate(skf.split(X, y), start=1):
        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
        y_train, y_test = y[train_idx], y[test_idx]

        for name, clf in models.items():
            preprocessor = _build_preprocessor(X_train)
            pipe = _build_pipeline(preprocessor, clf)
            pipe.fit(X_train, y_train)

            proba_test = pipe.predict_proba(X_test)[:, 1]
            proba_train = pipe.predict_proba(X_train)[:, 1]

            roc_test = roc_auc_score(y_test, proba_test)
            pr_test = average_precision_score(y_test, proba_test)
            brier_test = brier_score_loss(y_test, proba_test)

            roc_train = roc_auc_score(y_train, proba_train)
            pr_train = average_precision_score(y_train, proba_train)

            records.append({
                "model": name,
                "fold": fold_idx,
                "roc_auc_test": roc_test,
                "pr_auc_test": pr_test,
                "brier_test": brier_test,
                "roc_auc_train": roc_train,
                "pr_auc_train": pr_train,
            })

            print(f"  Fold {fold_idx} | {name:<25s} | "
                  f"ROC-AUC {roc_test:.4f} | PR-AUC {pr_test:.4f} | "
                  f"Brier {brier_test:.4f}")

    return pd.DataFrame(records)

# ------------------------------------------------------------------
# Reporting helpers
# ------------------------------------------------------------------
def print_summary_table(results: pd.DataFrame) -> None:
    """Aggregate per-model means and standard deviations."""
    print("\n" + "=" * 72)
    print("MODEL COMPARISON SUMMARY  (Stratified {}-Fold CV)".format(N_FOLDS))
    print("=" * 72)

    agg = results.groupby("model").agg(
        roc_auc_mean=("roc_auc_test", "mean"),
        roc_auc_std=("roc_auc_test", "std"),
        pr_auc_mean=("pr_auc_test", "mean"),
        pr_auc_std=("pr_auc_test", "std"),
        brier_mean=("brier_test", "mean"),
        brier_std=("brier_test", "std"),
    )

    for name, row in agg.iterrows():
        print(f"\n--- {name} ---")
        print(f"  ROC-AUC : {row.roc_auc_mean:.4f} ± {row.roc_auc_std:.4f}")
        print(f"  PR-AUC  : {row.pr_auc_mean:.4f} ± {row.pr_auc_std:.4f}")
        print(f"  Brier   : {row.brier_mean:.4f} ± {row.brier_std:.4f}  (lower is better)")

def print_stability_report(results: pd.DataFrame) -> None:
    """Report variance / coefficient of variation to demonstrate stability."""
    print("\n" + "=" * 72)
    print("STABILITY REPORT  (Performance Variance Across Folds)")
    print("=" * 72)

    for name, grp in results.groupby("model"):
        roc_cv = grp["roc_auc_test"].std() / grp["roc_auc_test"].mean() * 100
        pr_cv = grp["pr_auc_test"].std() / grp["pr_auc_test"].mean() * 100
        print(f"\n--- {name} ---")
        print(f"  ROC-AUC CV%: {roc_cv:.2f}%   (std / mean × 100)")
        print(f"  PR-AUC  CV%: {pr_cv:.2f}%")
        if roc_cv < 2.0:
            print("  → Very stable across folds")
        elif roc_cv < 5.0:
            print("  → Reasonably stable")
        else:
            print("  → High variance — consider more data or regularisation")

def print_overfitting_check(results: pd.DataFrame) -> None:
    """Compare train vs test metrics to flag overfitting."""
    print("\n" + "=" * 72)
    print("OVERFITTING CHECK  (Train vs. Test Gap)")
    print("=" * 72)

    agg = results.groupby("model").agg(
        roc_train=("roc_auc_train", "mean"),
        roc_test=("roc_auc_test", "mean"),
        pr_train=("pr_auc_train", "mean"),
        pr_test=("pr_auc_test", "mean"),
    )

    for name, row in agg.iterrows():
        roc_gap = row.roc_train - row.roc_test
        pr_gap = row.pr_train - row.pr_test
        print(f"\n--- {name} ---")
        print(f"  ROC-AUC  train: {row.roc_train:.4f}  test: {row.roc_test:.4f}  gap: {roc_gap:+.4f}")
        print(f"  PR-AUC   train: {row.pr_train:.4f}  test: {row.pr_test:.4f}  gap: {pr_gap:+.4f}")
        if roc_gap < 0.02:
            print("  → Minimal overfitting")
        elif roc_gap < 0.05:
            print("  → Moderate overfitting — monitor closely")
        else:
            print("  → Significant overfitting — consider stronger regularisation")

def print_calibration_report(
    X: pd.DataFrame,
    y: np.ndarray,
    train_idx: np.ndarray,
    test_idx: np.ndarray,
) -> None:
    """Fit each model once on a held-out split and print calibration curves."""
    print("\n" + "=" * 72)
    print("CALIBRATION ANALYSIS  (10-bin reliability diagram)")
    print("=" * 72)

    X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
    y_train, y_test = y[train_idx], y[test_idx]

    models = get_models()
    for name, clf in models.items():
        preprocessor = _build_preprocessor(X_train)
        pipe = _build_pipeline(preprocessor, clf)
        pipe.fit(X_train, y_train)

        proba = pipe.predict_proba(X_test)[:, 1]
        brier = brier_score_loss(y_test, proba)

        fraction_pos, mean_predicted = calibration_curve(
            y_test, proba, n_bins=10, strategy="uniform"
        )

        print(f"\n--- {name}  (Brier score: {brier:.4f}) ---")
        print(f"  {'Bin':>4s}  {'Mean predicted':>15s}  {'Fraction positive':>18s}  {'Deviation':>10s}")
        for i, (mp, fp) in enumerate(zip(mean_predicted, fraction_pos), start=1):
            dev = fp - mp
            print(f"  {i:>4d}  {mp:>15.4f}  {fp:>18.4f}  {dev:>+10.4f}")

        # Overall assessment
        max_dev = np.max(np.abs(fraction_pos - mean_predicted))
        if max_dev < 0.05:
            print("  → Well calibrated")
        elif max_dev < 0.10:
            print("  → Reasonably calibrated")
        else:
            print("  → Poor calibration — consider Platt scaling or isotonic regression")

# ------------------------------------------------------------------
# Main
# ------------------------------------------------------------------
def main() -> None:
    print("Loading and cleaning data …")
    df = load_and_clean(RESOLVED_DATA_PATH)
    X = df.drop(columns=["target"])
    y = df["target"].to_numpy()

    print(f"Dataset: {X.shape[0]:,} rows, {X.shape[1]} features")
    print(f"No-show rate: {y.mean():.2%}\n")

    # --- Stratified K-Fold CV ---
    print(f"Running {N_FOLDS}-fold Stratified Cross-Validation …\n")
    results = cross_validate_models(X, y)

    # --- Reports ---
    print_summary_table(results)
    print_stability_report(results)
    print_overfitting_check(results)

    # --- Calibration (using first fold split for demonstration) ---
    skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=RANDOM_STATE)
    first_train_idx, first_test_idx = next(iter(skf.split(X, y)))
    print_calibration_report(X, y, first_train_idx, first_test_idx)

    # --- Final recommendation ---
    print("\n" + "=" * 72)
    print("RECOMMENDATION")
    print("=" * 72)
    agg = results.groupby("model")["roc_auc_test"].mean()
    best_model = agg.idxmax()
    print(f"\nBest model by mean ROC-AUC: {best_model} ({agg[best_model]:.4f})")

    # Check if improvement over baseline is meaningful
    lr_auc = agg.get("Logistic Regression", 0)
    best_auc = agg[best_model]
    if best_model != "Logistic Regression" and (best_auc - lr_auc) < 0.005:
        print("  → Improvement over Logistic Regression is marginal (<0.5%).")
        print("  → Logistic Regression may still be preferred for interpretability.")
    elif best_model == "Logistic Regression":
        print("  → Baseline model is already the strongest — no need for complexity.")
    else:
        improvement = (best_auc - lr_auc) * 100
        print(f"  → Improvement of {improvement:.2f}% ROC-AUC over Logistic Regression.")
        print(f"  → Consider deploying {best_model} if interpretability is not critical.")

    print()

if __name__ == "__main__":
    main()

Output:
Loading and cleaning data …
Dataset: 110,526 rows, 13 features
No-show rate: 20.19%

Running 5-fold Stratified Cross-Validation …

  Fold 1 | Logistic Regression       | ROC-AUC 0.6705 | PR-AUC 0.3028 | Brier 0.2308
  Fold 1 | Random Forest             | ROC-AUC 0.7348 | PR-AUC 0.3628 | Brier 0.2114
  Fold 1 | Gradient Boosting         | ROC-AUC 0.7437 | PR-AUC 0.3850 | Brier 0.1418
  Fold 2 | Logistic Regression       | ROC-AUC 0.6640 | PR-AUC 0.3051 | Brier 0.2307
  Fold 2 | Random Forest             | ROC-AUC 0.7260 | PR-AUC 0.3533 | Brier 0.2118
  Fold 2 | Gradient Boosting         | ROC-AUC 0.7396 | PR-AUC 0.3805 | Brier 0.1425
  Fold 3 | Logistic Regression       | ROC-AUC 0.6722 | PR-AUC 0.3062 | Brier 0.2307
  Fold 3 | Random Forest             | ROC-AUC 0.7247 | PR-AUC 0.3512 | Brier 0.2130
  Fold 3 | Gradient Boosting         | ROC-AUC 0.7398 | PR-AUC 0.3787 | Brier 0.1425
  Fold 4 | Logistic Regression       | ROC-AUC 0.6636 | PR-AUC 0.3000 | Brier 0.2310
  Fold 4 | Random Forest             | ROC-AUC 0.7294 | PR-AUC 0.3568 | Brier 0.2118
  Fold 4 | Gradient Boosting         | ROC-AUC 0.7422 | PR-AUC 0.3805 | Brier 0.1424
  Fold 5 | Logistic Regression       | ROC-AUC 0.6661 | PR-AUC 0.3056 | Brier 0.2307
  Fold 5 | Random Forest             | ROC-AUC 0.7245 | PR-AUC 0.3544 | Brier 0.2115
  Fold 5 | Gradient Boosting         | ROC-AUC 0.7388 | PR-AUC 0.3771 | Brier 0.1426

========================================================================
MODEL COMPARISON SUMMARY  (Stratified 5-Fold CV)
========================================================================

--- Gradient Boosting ---
  ROC-AUC : 0.7408 ± 0.0020
  PR-AUC  : 0.3804 ± 0.0029
  Brier   : 0.1424 ± 0.0003  (lower is better)

--- Logistic Regression ---
  ROC-AUC : 0.6673 ± 0.0039
  PR-AUC  : 0.3040 ± 0.0026
  Brier   : 0.2308 ± 0.0002  (lower is better)

--- Random Forest ---
  ROC-AUC : 0.7279 ± 0.0043
  PR-AUC  : 0.3557 ± 0.0044
  Brier   : 0.2119 ± 0.0006  (lower is better)

========================================================================
STABILITY REPORT  (Performance Variance Across Folds)
========================================================================

--- Gradient Boosting ---
  ROC-AUC CV%: 0.28%   (std / mean × 100)
  PR-AUC  CV%: 0.77%
  → Very stable across folds

--- Logistic Regression ---
  ROC-AUC CV%: 0.58%   (std / mean × 100)
  PR-AUC  CV%: 0.84%
  → Very stable across folds

--- Random Forest ---
  ROC-AUC CV%: 0.60%   (std / mean × 100)
  PR-AUC  CV%: 1.24%
  → Very stable across folds

========================================================================
OVERFITTING CHECK  (Train vs. Test Gap)
========================================================================

--- Gradient Boosting ---
  ROC-AUC  train: 0.7696  test: 0.7408  gap: +0.0288
  PR-AUC   train: 0.4454  test: 0.3804  gap: +0.0650
  → Moderate overfitting — monitor closely

--- Logistic Regression ---
  ROC-AUC  train: 0.6710  test: 0.6673  gap: +0.0037
  PR-AUC   train: 0.3062  test: 0.3040  gap: +0.0022
  → Minimal overfitting

--- Random Forest ---
  ROC-AUC  train: 0.7458  test: 0.7279  gap: +0.0179
  PR-AUC   train: 0.3872  test: 0.3557  gap: +0.0315
  → Minimal overfitting

========================================================================
CALIBRATION ANALYSIS  (10-bin reliability diagram)
========================================================================

--- Logistic Regression  (Brier score: 0.2308) ---
   Bin   Mean predicted   Fraction positive   Deviation
     1           0.1786              0.1818     +0.0032
     2           0.2789              0.0721     -0.2068
     3           0.3561              0.0974     -0.2587
     4           0.4460              0.1799     -0.2660
     5           0.5479              0.2893     -0.2587
     6           0.6445              0.3448     -0.2997
     7           0.7424              0.3383     -0.4041
     8           0.8377              0.2983     -0.5394
     9           0.9333              0.3333     -0.6000
  → Poor calibration — consider Platt scaling or isotonic regression

--- Random Forest  (Brier score: 0.2114) ---
   Bin   Mean predicted   Fraction positive   Deviation
     1           0.1894              0.0163     -0.1731
     2           0.2378              0.0361     -0.2017
     3           0.3281              0.1853     -0.1428
     4           0.4843              0.1799     -0.3044
     5           0.5512              0.2437     -0.3075
     6           0.6278              0.3695     -0.2583
     7           0.7119              0.5714     -0.1404
  → Poor calibration — consider Platt scaling or isotonic regression

--- Gradient Boosting  (Brier score: 0.1418) ---
   Bin   Mean predicted   Fraction positive   Deviation
     1           0.0417              0.0321     -0.0096
     2           0.1675              0.1811     +0.0136
     3           0.2497              0.2536     +0.0038
     4           0.3437              0.3376     -0.0061
     5           0.4361              0.4521     +0.0159
     6           0.5376              0.5714     +0.0339
     7           0.6441              0.5882     -0.0558
     8           0.7009              1.0000     +0.2991
  → Poor calibration — consider Platt scaling or isotonic regression

========================================================================
RECOMMENDATION
========================================================================

Best model by mean ROC-AUC: Gradient Boosting (0.7408)
  → Improvement of 7.35% ROC-AUC over Logistic Regression.
  → Consider deploying Gradient Boosting if interpretability is not critical.
